meilisearch-lib/src/document_formats.rs

```diff
@@ -0,0 +1,364 @@
+        builder.add_documents(&value)?;
+    }
+
+    builder.finish()?;
+
+    Ok(())
+}
+
+/// read json from input and write an obkv batch to writer.
+pub fn read_json(input: impl Read, writer: impl Write + Seek) -> Result<()> {
+    let mut builder = DocumentBatchBuilder::new(writer).unwrap();
+
+    let documents: Vec<Map<String, Value>> =
+        malformed!(PayloadType::Json, serde_json::from_reader(input))?;
+    builder.add_documents(documents).unwrap();
+    builder.finish().unwrap();
+
+    Ok(())
+}
+
+enum AllowedType {
+    String,
+    Number,
+}
+
+fn parse_csv_header(header: &str) -> (String, AllowedType) {
+    // if there are several separators we only split on the last one.
+    match header.rsplit_once(':') {
+        Some((field_name, field_type)) => match field_type {
+            "string" => (field_name.to_string(), AllowedType::String),
+            "number" => (field_name.to_string(), AllowedType::Number),
+            // if the pattern isn't reconized, we keep the whole field.
+            _otherwise => (header.to_string(), AllowedType::String),
+        },
+        None => (header.to_string(), AllowedType::String),
+    }
+}
+
+pub struct CsvDocumentIter<R>
+where
+    R: Read,
+{
+    documents: StringRecordsIntoIter<R>,
+    headers: Vec<(String, AllowedType)>,
+}
+
+impl<R: Read> CsvDocumentIter<R> {
+    pub fn from_reader(reader: R) -> IoResult<Self> {
+        let mut records = CsvReader::from_reader(reader);
+
+        let headers = records
+            .headers()?
+            .into_iter()
+            .map(parse_csv_header)
+            .collect();
+
+        Ok(Self {
+            documents: records.into_records(),
+            headers,
+        })
+    }
+}
+
+impl<R: Read> Iterator for CsvDocumentIter<R> {
+    type Item = Result<Map<String, Value>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let csv_document = self.documents.next()?;
+
+        match csv_document {
+            Ok(csv_document) => {
+                let mut document = Map::new();
+
+                for ((field_name, field_type), value) in
+                    self.headers.iter().zip(csv_document.into_iter())
+                {
+                    let parsed_value = match field_type {
+                        AllowedType::Number => {
+                            malformed!(PayloadType::Csv, value.parse::<f64>().map(Value::from))
+                        }
+                        AllowedType::String => Ok(Value::String(value.to_string())),
+                    };
+
+                    match parsed_value {
+                        Ok(value) => drop(document.insert(field_name.to_string(), value)),
+                        Err(e) => return Some(Err(e)),
+                    }
+                }
+
+                Some(Ok(document))
+            }
+            Err(e) => Some(Err(DocumentFormatError::MalformedPayload(
+                Box::new(e),
+                PayloadType::Csv,
+            ))),
+        }
+    }
+}
```

Comment:
Do you think it would be a good idea for another PR to move all this code to a separate crate?
So we can use it in `milli/http-ui` and `milli/benchmarks` instead of copy-pasting it everywhere

Reply:
I think, ultimately, the whole document format should be moved to another crate